pkg/cosesign1/check.go

```diff
@@ -0,0 +1,202 @@
+package cosesign1
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"log"
+	"reflect"
+
+	"github.com/veraison/go-cose"
+)
+
+type UnpackedCoseSign1 struct {
+	Issuer      string
+	Feed        string
+	ContentType string
+	Pubkey      string
+	Pubcert     string
+	Payload     []byte
+	CertChain   []*x509.Certificate
+}
+
+/*
+	Create a pem of the form:
+
+-----BEGIN CERTIFICATE-----
+single line base64 standard encoded raw DER certificate
+-----END CERTIFICATE-----
+
+	Note that there are no extra line breaks added and that a string compare will need to accomodate that.
+*/
+
+func UnpackAndValidateCOSE1CertChain(raw []byte, optionaPubKeyPEM []byte, requireKNownAuthority bool, verbose bool) (UnpackedCoseSign1, error) {
+	var msg cose.Sign1Message
+	err := msg.UnmarshalCBOR(raw)
+	if err != nil {
+		return UnpackedCoseSign1{}, err
+	}
+
+	protected := msg.Headers.Protected
+	algo := protected[cose.HeaderLabelAlgorithm]
+
+	if verbose {
+		log.Printf("algo %d aka %s", algo.(cose.Algorithm), algo.(cose.Algorithm))
+	}
+
+	x5RawChain, x5RawChainPresent := protected[cose.HeaderLabelX5Chain] // The spec says this is ordered - leaf, intermediates, root. X5Bag is unordered and woould need sorting
+	if !x5RawChainPresent {
+		return UnpackedCoseSign1{}, fmt.Errorf("x5Chain missing")
+	}
+
+	var issuer string
+	val, valPresent := protected[HeaderLabelIssuer]
+	if valPresent {
+		issuer = val.(string)
```

Comment:
are we guaranteed val (here and below) are strings?

Reply:
ok. I have protected these and the chain array.